bundler/lib/bundler/source/rubygems.rb

```diff
@@ -506,6 +506,11 @@ def extension_cache_slug(spec)
         return unless remote = spec.remote
         remote.cache_slug
       end
+
+      def fetcher_for_remote(remote)
+        fetchers.find {|f| f.remote?(remote) } ||
+          raise(ArgumentError, "No remote found")
+      end
```

Comment:
It felt a bit off having to expose this `Fetcher#remote?` method just for this. How about

```diff
diff --git a/bundler/lib/bundler/fetcher.rb b/bundler/lib/bundler/fetcher.rb
index 844a3fa94..46ee68fed 100644
--- a/bundler/lib/bundler/fetcher.rb
+++ b/bundler/lib/bundler/fetcher.rb
@@ -214,10 +214,6 @@ def gem_remote_fetcher
       end
     end
 
-    def remote?(remote)
-      @remote == remote
-    end
-
     private
 
     def available_fetchers
diff --git a/bundler/lib/bundler/source/rubygems.rb b/bundler/lib/bundler/source/rubygems.rb
index e72369be8..428ddad24 100644
--- a/bundler/lib/bundler/source/rubygems.rb
+++ b/bundler/lib/bundler/source/rubygems.rb
@@ -255,11 +255,16 @@ def unmet_deps
         end
       end
 
-      def fetchers
-        @fetchers ||= remotes.map do |uri|
+      def remote_fetchers
+        @remote_fetchers ||= remotes.map do |uri|
           remote = Source::Rubygems::Remote.new(uri)
-          Bundler::Fetcher.new(remote)
-        end
+
+          [remote, Bundler::Fetcher.new(remote)]
+        end.to_h
+      end
+
+      def fetchers
+        @fetchers ||= remote_fetchers.values
       end
 
       def double_check_for(unmet_dependency_names)
@@ -480,7 +485,7 @@ def cache_path
       def download_gem(spec, download_cache_path, previous_spec = nil)
         uri = spec.remote.uri
         Bundler.ui.confirm("Fetching #{version_message(spec, previous_spec)}")
-        Bundler.rubygems.download_gem(spec, uri, download_cache_path, fetcher_for_remote(spec.remote).gem_remote_fetcher)
+        Bundler.rubygems.download_gem(spec, uri, download_cache_path, remote_fetchers[spec.remote].gem_remote_fetcher)
       end
 
       # Returns the global cache path of the calling Rubygems::Source object.
@@ -506,11 +511,6 @@ def extension_cache_slug(spec)
         return unless remote = spec.remote
         remote.cache_slug
       end
-
-      def fetcher_for_remote(remote)
-        fetchers.find {|f| f.remote?(remote) } ||
-          raise(ArgumentError, "No remote found")
-      end
     end
   end
 end
```

Reply:
I second this suggestion. The current code is basically using an array as a hash, so using an actual hash should simplify the code considerably.

Reply:
This was fixed, thanks @segiddins!